src/auth/auth.service.ts

```diff
@@ -0,0 +1,276 @@
+    const { email, password } = dto;
+
+    const foundUser = await this.userModel.findOne({
+      where: {
+        email,
+      },
+    });
+
+    if (!foundUser) {
+      throw new BadRequestException('Wrong credentials');
+    }
+
+    const compareSuccess = await this.comparePasswords({
+      password,
+      hash: foundUser.hashedPassword,
+    });
+
+    if (!compareSuccess) {
+      throw new BadRequestException('Wrong credentials');
+    }
+
+    const tokens = await this.getTokens(foundUser.id, foundUser.email);
+    await this.updateRtHash(foundUser.id, tokens.refresh_token);
+
+    return tokens;
+  }
+
+  async signout(id: typeof randomUUID): Promise<boolean> {
+    const data = {
+      hashedRT: null,
+    };
+    await this.userModel
+      .update(data, {
+        where: {
+          id: id,
+        },
+      })
+      .then((result) => {
+        console.log(result);
+        return true;
+      })
+      .catch((err) => {
+        console.log('user dosent exist');
+        console.error(err);
+        return false;
+      });
+    return true;
+  }
+
+  async verifyAccessToken(token: string): Promise<{ expired: boolean; payload: any }> {
+    return new Promise((resolve, reject) => {
+      jwt.verify(token, 'at-secret', (err, decoded) => {
+        if (err) {
+          if (err.name === 'TokenExpiredError') {
+            //const timeRemaining = Math.floor((err.expiredAt.getTime() - Date.now()) / 1000);
+            resolve({ expired: true, payload: null });
+          } else {
+            reject(err);
+          }
+        } else {
+          resolve({ expired: false, payload: decoded });
+        }
+      });
+    });
+  }
+
+  async refreshTokens(id: typeof randomUUID, rt: string): Promise<Tokens> {
+    const user = await this.userModel.findOne({
+      where: {
+        id: id,
+      },
+    });
+    if (!user || !user.hashedRT) {
+      throw new ForbiddenException('Access Denied');
+    }
+
+    const rtMatches = await this.comparePasswords({ password: rt, hash: user.hashedRT });
+    if (!rtMatches) {
+      throw new ForbiddenException('Access Denied');
+    }
+
+    const tokens = await this.getTokens(user.id, user.email);
+    await this.updateRtHash(user.id, tokens.refresh_token);
+
+    return tokens;
+  }
+
+  async updateRtHash(id: typeof randomUUID, rt: string): Promise<void> {
+    const hash = await this.hashPassword(rt);
+    await this.userModel.update(
+      {
+        hashedRT: hash,
+      },
+      {
+        where: {
+          id: id,
+        },
+      },
+    );
+  }
+
+  async getTokens(id: typeof randomUUID, email: string): Promise<Tokens> {
+    const jwtPayload: JwtPayload = {
+      sub: id,
+      email: email,
+    };
+
+    const [at, rt] = await Promise.all([
+      this.jwtService.signAsync(jwtPayload, {
+        secret: this.config.get<string>('AT_SECRET'),
+        expiresIn: '7d',
+      }),
+      this.jwtService.signAsync(jwtPayload, {
```

Comment:
refresh token is not a jwt, its just a random string

Reply:
will it be okay if we keep it a jwt , cause we wrote the whole logic considering that , and it is part of many routes